Add: 3D spatial audio positioning

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Adds immersive 3D audio positioning for MUD games, allowing sounds to be placed around the player in virtual space for enhanced gameplay experiences.

**Key Features:**
- Position sounds anywhere around the player using azimuth/elevation/distance coordinates
- Complete Lua API: [`playSpatialSound()`](https://wiki.mudlet.org/w/Area_51#playSpatialSound.2C_PR_.238452), [`updateSpatialSound()`](https://wiki.mudlet.org/w/Area_51#updateSpatialSound.2C_PR_.238452), [`stopSpatialSound()`](https://wiki.mudlet.org/w/Area_51#stopSpatialSound.2C_PR_.238452), [`pauseSpatialSound`](https://wiki.mudlet.org/w/Area_51#pauseSpatialSound.2C_PR_.238452), [`removeSpatialSound`](https://wiki.mudlet.org/w/Area_51#removeSpatialSound.2C_PR_.238452), [`getSpatialSounds`](https://wiki.mudlet.org/w/Area_51#getSpatialSounds.2C_PR_.238452), [`setSpatialListener()`](https://wiki.mudlet.org/w/Area_51#setSpatialListener.2C_PR_.238452), [`setSpatialMasterVolume()`](https://wiki.mudlet.org/w/Area_51#setSpatialMasterVolume.2C_PR_.238452), [`playSpatialTestTone()`](https://wiki.mudlet.org/w/Area_51#playSpatialTestTone.2C_PR_.238452)
- Server-driven spatial audio via new GMCP [`Client.Media.Spatial`](https://wiki.mudlet.org/w/Area_51#Client.Media.Spatial.2C_PR_.238452) for the Mud Client Media Protocol (MCMP)
- Room acoustics simulation with materials, reverb, and occlusion effects
- Seven built-in audio tests for speaker setup and calibration
- Independent mute controls for player scripts vs server sounds
- Automatic file downloading and sound generation capabilities

#### Motivation for adding to Mudlet

**Addresses Long-Standing Community Requests:**
- **Issue #7014**: Implements stereo panning capabilities using Qt6 SpatialAudio framework
- **Issue #3600**: Extends MCMP with spatial positioning beyond basic audio playback

**User Benefits:**
- Creates immersive gaming experiences where combat sounds come from the right direction
- Enables atmospheric audio like footsteps approaching from behind or spells casting nearby  
- Supports both simple left/right panning and full 3D positioning
- Game servers can create rich audio environments without requiring custom client modifications
- Maintains full compatibility with existing Mudlet audio functions

#### Other info (issues closed, discussion etc)

**Technical Foundation:**
- Built on Qt6 SpatialAudio for professional-quality audio positioning
- Follows established Mudlet patterns similar to existing media system
- Uses intuitive spherical coordinates (azimuth/elevation/distance) for positioning
- Separates user API calls from server messages for proper mute handling

**Professional Testing Suite:**
Accessible via Settings → Special Options for validating audio setup:
1. White Noise - Circular Movement (verify basic positioning works)
2. Pink Noise - Speaker Calibration (industry standard audio reference)
3. Sine Wave - Frequency Test (test different tones at various positions)
4. 7.1 Surround - Professional Setup (complete surround sound calibration)
5. Vertical Positioning Test (test sounds above and below ear level)
6. Distance Perception Test (verify depth and attenuation effects)
7. Room Materials & Occlusion Test (test acoustic environment simulation)

Resolves #7014, Resolves #3600

<img width="1070" height="326" alt="Screenshot 2025-10-26 at 12 34 45 PM" src="https://github.com/user-attachments/assets/2733cd24-6d45-4170-b656-2cdfb5dc6455" />

---

Turn Sound On!

https://github.com/user-attachments/assets/0e719dcf-8642-4b9a-b434-157da464f3ee

mpconley committed:
Implement comprehensive spatial audio system with Qt6 SpatialAudio

- Add TSpatialAudio engine with full 3D positioning support
- Implement playSpatialSound(), updateSpatialSound(), stopSpatialSound() Lua API
- Support spherical coordinates (azimuth/elevation/distance) positioning
- Add GMCP Client.Media.Spatial protocol for server-driven spatial audio
- Include room acoustics with materials, reverb, and occlusion effects
- Add comprehensive spatial audio test suite in Settings → Special Options
- Support both API and GMCP protocol sources with separate mute controls
- Enable file downloading and caching for spatial audio assets
- Generate test tones (white/pink noise, sine waves) for calibration

Addresses GitHub issues #7014 (stereo panning) and #3600 (MCMP enhancement)
by providing Qt6 SpatialAudio-based solution that supports both basic stereo
panning and full 3D spatial positioning beyond original MCMP specification.

mpconley committed:
Fix code style issues identified by CodeFactor

- Mark TSpatialAudioRoom single-parameter constructor as explicit
- Fix parenthesis formatting in parseJSONByRoom dimensions parsing

mpconley committed:
Merge branch 'development' into media_pan

mpconley commented:
Ready for review - documented, and test out your system in Special Options

mpconley commented:
Note: Purposefully separated from `TMedia` and `playSoundFile()`. I tried an integration, but this is different. Upside is no new bugs for the standard media platform. The only integration is the Mute button, where it will also stop spatial sounds.

mpconley committed:
Merge branch 'development' into media_pan

mpconley committed:
Remove duplicate WallMaterial enum, use Qt's QAudioRoom::Material

- Eliminates code duplication by removing custom WallMaterial enum
- Updates TSpatialAudio::setWallMaterial to use QAudioRoom::Material directly
- Removes unsafe static_cast operations in material handling
- Updates all references across TSpatialAudio, TLuaInterpreterSpatialAudio, and dlgProfilePreferences
- Improves code safety and maintainability for spatial audio features

SlySven commented on src/TLuaInterpreterSpatialAudio.cpp:
```diff
@@ -0,0 +1,652 @@
+    // Optional: position {azimuth, elevation, distance}
+    lua_getfield(L, 1, "position");
+
+    if (lua_istable(L, -1)) {
```
Am I right in thinking this is a static setting - there is no provision to change the sound location during playback - e.g. with the sound of an arrow flying from a tower 45 degrees above and to the right of the listener down to the ground a meter to the left and behind the listener (or even into their middle of their chest)?

SlySven replied:
TBH I think you need a way to describe a 3D path for the sound verses time...

mpconley replied:
Check out [updateSpatialSound](https://wiki.mudlet.org/w/Area_51#updateSpatialSound.2C_PR_.238452).

mpconley replied:
```
// Enhanced positioning with velocity
Client.Media.Spatial.Play {
  "key": "moving_creature",
  "position": [45, 0, 5],
  "velocity": [2.0, 0, 0],  // m/s for Doppler
  "trajectory": {
    "path": [[45,0,5], [90,0,3], [135,0,2]],
    "duration": 5.0,
    "easing": "smooth"
  }
}
```
or
```
-- Enhanced positioning
setSpatialSourceTrajectory("ambient_wind", {
  positions = {{0,0,5}, {10,0,5}, {0,0,5}},
  duration = 30,
  loop = true,
  easing = "sine"
})
```
Could be some examples.

It may be nice to add more features for Mudlet to manage. Some of this could be handled in scripts crafted by users with the existing APIs, too.

I'd like to get the basics right, and then adding in features could come along later.

SlySven commented on src/TLuaInterpreterSpatialAudio.cpp:
```diff
@@ -0,0 +1,652 @@
+                QAudioRoom::Material mat = QAudioRoom::Material::BrickBare;
+                
+                // Map material name to enum
+                if (material == "brick") {
```
What is the way to specify an open area with no walls - i.e. out in the open?

mpconley committed:
Fix: Spatial audio privacy and protocol separation

- Remove output mode sharing in GMCP settings response (privacy fix)
- Restrict GMCP status response to server-managed sources only
- Add missing QAudioRoom::Material enum cases (ParquetOnConcrete, WaterOrIceSurface, UniformMaterial)
- Complete materialToString() switch statement to resolve compiler warnings
- Maintain proper separation between API and GMCP protocol sources
- Ensure game servers only know about spatial audio sources they created

mpconley reviewed (commented):
Thank you for the great question about open areas! 

For open areas with no walls, use the `"transparent"` material (or `"air"` as an alias). This maps to Qt's `QAudioRoom::Material::Transparent` which provides no reflections.

For best open-area simulation:
```lua
playSpatialSound({
    key = "outdoor_sound",
    name = "wind.wav", 
    room = {
        dimensions = {100, 100, 50},  -- Very large space
        material = "transparent",     -- No wall reflections  
        reverb = 0.0,                -- No reverb
        reflection = 0.0             -- No reflections
    }
})
```

The implementation already supports this - I've added a clarifying comment to the `stringToMaterial` function in TSpatialAudio.cpp to make this more discoverable.

mpconley committed:
Docs: Add clarifying comment for open area spatial audio

Added comment to stringToMaterial() function to clarify that 'transparent'
or 'air' materials should be used for open areas with no walls/reflections.
This addresses SlySven's review question about outdoor environments.

mpconley committed:
Fix: Remove invalid GMCP module from Core.Supports.Set

Removed 'Client.Media.Spatial 1' from Core.Supports.Set as GMCP
standards only allow two nodes in module specifications. Spatial
audio capabilities are properly covered under 'Client.Media 1'.

mpconley committed:
Fix spatial audio shutdown crash using QScopedPointer pattern

- Changed TSpatialAudio from raw pointer to QScopedPointer in Host class
- Established Qt parent-child relationship for automatic cleanup
- Simplified destructor to use Qt's automatic object destruction
- Updated all references to work with QScopedPointer access pattern
- Follows proven TMedia design pattern for lifecycle management
- Fixes AddressSanitizer crash in QAudioEnginePrivate::removeSpatialSound()

mpconley committed:
Refactor: Use centralized stringToMaterial() helper in Lua API

- Replaced hardcoded if-else material mapping with TSpatialAudio::stringToMaterial()
- Ensures consistency between GMCP and Lua API material handling
- Provides access to all 23 Qt6 spatial audio materials instead of just 4
- Improves maintainability by centralizing material mapping logic
- Follows DRY principle by reusing existing comprehensive helper function

mpconley committed:
Cleanup: Remove whitespace-only lines from spatial audio files

- Cleaned up 79 whitespace-only lines in TLuaInterpreterSpatialAudio.cpp
- Cleaned up 136 whitespace-only lines in TSpatialAudio.cpp
- Cleaned up 34 whitespace-only lines in TSpatialAudio.h
- Preserved newline characters to maintain code structure and readability
- Total: 249 whitespace-only lines removed across all spatial audio files
- Improves code quality and consistency with project standards

mpconley committed:
Format: Improve code readability in TSpatialAudio.cpp

- Add strategic blank lines to separate logical code sections
- Remove unnecessary blank lines for cleaner flow
- Improve visual separation around conditional blocks and loops
- Maintain consistent spacing throughout the file

mpconley committed:
Refactor: Remove 'drywall' alias from spatial audio materials

Removed the 'drywall' alias for sheetrock material as it's anachronistic
for MUD/fantasy game contexts. The stringToMaterial() function now only
accepts 'sheetrock' for this material type, making the API more thematically
appropriate for typical MUD environments.

mpconley committed:
Add: 'stone' alias for sheetrock material in spatial audio

Added 'stone' as an alias for the sheetrock material to provide
a more thematically appropriate term for fantasy MUD environments
where 'sheetrock' may be anachronistic.

mpconley committed:
Merge branch 'development' into media_pan

mpconley commented:
Since this is so different, a thought is this should be `Client.Spatial` instead of wired into `Client.Media`.

mpconley committed:
Merge branch 'development' into media_pan

mpconley committed:
Merge branch 'development' into media_pan

mpconley committed:
Merge branch 'development' into media_pan

mpconley committed:
Merge branch 'development' into media_pan

mpconley commented:
OK foundation, cool stuff, will rework this in a future PR, thinking a little broader.
